Mapping fails with 2 Types with fluent, chained accessors

hi!

we are using lombok for everything and i discovered a problem when mapping back and forth between two types that use fluent and chained accessors.

the code example below adds 2 `PropertyMap`s, one for Dto -> Entity and one for the other way. there is just 1 mapped property which is mapped explicitly. adding either one of the `PropertyMap`s works while adding both results in the exception below.

is there any nice way to solve this?

```
Exception in thread "main" org.modelmapper.ConfigurationException: ModelMapper configuration errors:

1) Failed to configure mappings

1 error
	at org.modelmapper.internal.Errors.throwConfigurationExceptionIfErrorsExist(Errors.java:241)
	at org.modelmapper.internal.ExplicitMappingBuilder.build(ExplicitMappingBuilder.java:207)
	at org.modelmapper.internal.TypeMapImpl.addMappings(TypeMapImpl.java:73)
	at org.modelmapper.internal.TypeMapStore.getOrCreate(TypeMapStore.java:101)
	at org.modelmapper.ModelMapper.addMappings(ModelMapper.java:93)
	at io.sourcy.playground.Main.main(Main.java:21)
Caused by: java.lang.ClassCastException: org.modelmapper.internal.PropertyInfoImpl$MethodMutator cannot be cast to org.modelmapper.internal.Accessor
	at org.modelmapper.internal.PropertyInfoRegistry.accessorFor(PropertyInfoRegistry.java:61)
	at org.modelmapper.internal.ExplicitMappingVisitor$MappingCapturingVisitor.recordSourceMethod(ExplicitMappingVisitor.java:256)
	at org.modelmapper.internal.ExplicitMappingVisitor$MappingCapturingVisitor.visitEnd(ExplicitMappingVisitor.java:204)
	at org.modelmapper.internal.asm.ClassReader.b(Unknown Source)
	at org.modelmapper.internal.asm.ClassReader.accept(Unknown Source)
	at org.modelmapper.internal.asm.ClassReader.accept(Unknown Source)
	at org.modelmapper.internal.ExplicitMappingBuilder.visitPropertyMap(ExplicitMappingBuilder.java:222)
	at org.modelmapper.PropertyMap.configure(PropertyMap.java:380)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.modelmapper.internal.ExplicitMappingBuilder.build(ExplicitMappingBuilder.java:195)
	... 4 more
```


```package io.sourcy.playground;


import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;
import org.modelmapper.ModelMapper;
import org.modelmapper.PropertyMap;

public final class Main {
    public static void main(String[] args) {
        final ModelMapper mm = new ModelMapper();
        mm.addMappings(new PropertyMap<SomeDto, SomeEntity>() {
            @Override
            protected void configure() {
                map(source.someValue(), destination.someValueWithDifferentName(null));
            }
        });

        mm.addMappings(new PropertyMap<SomeEntity, SomeDto>() {
            @Override
            protected void configure() {
                map(source.someValueWithDifferentName(), destination.someValue(null));
            }
        });
//
//
//        final SomeDto dto = new SomeDto("hello");
//        final SomeEntity entity = mm.map(dto, SomeEntity.class);
//        System.err.println(entity);
//        final SomeDto newDto = mm.map(entity, SomeDto.class);
//        System.err.println(newDto);
    }

    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    @Accessors(fluent = true, chain = true)
    public static class SomeDto {
        private String someValue;
    }

    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    @Accessors(fluent = true, chain = true)
    public static class SomeEntity {
        private String someValueWithDifferentName;
    }
}
```

--- a/core/src/main/java/org/modelmapper/internal/ExplicitMappingVisitor.java
+++ b/core/src/main/java/org/modelmapper/internal/ExplicitMappingVisitor.java
@@ -8,6 +8,7 @@ import java.util.ArrayList;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
+import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 import org.modelmapper.internal.PropertyInfoImpl.FieldPropertyInfo;
@@ -45,7 +46,8 @@ public class ExplicitMappingVisitor extends ClassVisitor {
   private static final String MAP_SOURCE_METHOD_DESC = "(Ljava/lang/Object;)Ljava/lang/Object;";
   private static final String SKIP_DEST_METHOD_DESC = "(Ljava/lang/Object;)V";
   private static final String MAP_BOTH_METHOD_DESC = "(Ljava/lang/Object;Ljava/lang/Object;)V";
-  private static final String MUTATOR_METHOD_DESC_PATTERN = "(.+)V";
+
+  private static final Pattern MUTATOR_METHOD_DESC_PATTERN = Pattern.compile("^\\(.+\\)(.+)$");
 
   private final Errors errors;
   private final InheritingConfiguration config;
@@ -236,7 +238,13 @@ public class ExplicitMappingVisitor extends ClassVisitor {
     }
 
     private boolean isMutator(MethodInsnNode mn) {
-      return Pattern.matches(MUTATOR_METHOD_DESC_PATTERN, mn.desc);
+      Matcher matcher = MUTATOR_METHOD_DESC_PATTERN.matcher(mn.desc);
+      if (!matcher.find())
+        return false;
+
+      String returnType = matcher.group(1);
+      return returnType.equals("V")
+          || returnType.contains(mn.owner);
     }
 
     private void recordProperties() {

--- a/core/src/main/java/org/modelmapper/internal/PropertyInfoRegistry.java
+++ b/core/src/main/java/org/modelmapper/internal/PropertyInfoRegistry.java
@@ -24,7 +24,6 @@ import org.modelmapper.config.Configuration;
 import org.modelmapper.internal.PropertyInfoImpl.FieldPropertyInfo;
 import org.modelmapper.internal.PropertyInfoImpl.MethodAccessor;
 import org.modelmapper.internal.PropertyInfoImpl.MethodMutator;
-import org.modelmapper.spi.PropertyInfo;
 
 /**
  * Statically stores and retrieves MemberInfo by member and configuration. This registry is designed
@@ -34,21 +33,26 @@ import org.modelmapper.spi.PropertyInfo;
  * @author Jonathan Halterman
  */
 class PropertyInfoRegistry {
-  private static final Map<Integer, PropertyInfo> cache = new ConcurrentHashMap<Integer, PropertyInfo>();
+  private static final Map<Integer, Mutator> MUTATOR_CACHE = new ConcurrentHashMap<Integer, Mutator>();
+  private static final Map<Integer, Accessor> ACCESSOR_CACHE = new ConcurrentHashMap<Integer, Accessor>();
+  private static final Map<Integer, FieldPropertyInfo> FIELD_CACHE = new ConcurrentHashMap<Integer, FieldPropertyInfo>();
 
   private static Integer hashCodeFor(Class<?> initialType, String propertyName,
       Configuration configuration) {
     int result = 31 + initialType.hashCode();
     result = 31 * result + propertyName.hashCode();
     result = 31 * result + configuration.hashCode();
-    return Integer.valueOf(result);
+    return result;
   }
 
   /**
    * Returns an accessor for the {@code accessorName}, else {@code null} if none exists.
    */
   static Accessor accessorFor(Class<?> type, String accessorName, Configuration configuration) {
-    return (Accessor) cache.get(hashCodeFor(type, accessorName, configuration));
+    Integer hashCode = hashCodeFor(type, accessorName, configuration);
+    if (ACCESSOR_CACHE.containsKey(hashCode))
+      return ACCESSOR_CACHE.get(hashCode);
+    return FIELD_CACHE.get(hashCode);
   }
 
   /**
@@ -58,10 +62,10 @@ class PropertyInfoRegistry {
   static synchronized Accessor accessorFor(Class<?> type, Method method,
       Configuration configuration, String name) {
     Integer hashCode = hashCodeFor(type, method.getName(), configuration);
-    Accessor accessor = (Accessor) cache.get(hashCode);
+    Accessor accessor = ACCESSOR_CACHE.get(hashCode);
     if (accessor == null) {
       accessor = new MethodAccessor(type, method, name);
-      cache.put(hashCode, accessor);
+      ACCESSOR_CACHE.put(hashCode, accessor);
     }
 
     return accessor;
@@ -73,10 +77,10 @@ class PropertyInfoRegistry {
   static synchronized FieldPropertyInfo fieldPropertyFor(Class<?> type, Field field,
       Configuration configuration, String name) {
     Integer hashCode = hashCodeFor(type, field.getName(), configuration);
-    FieldPropertyInfo fieldPropertyInfo = (FieldPropertyInfo) cache.get(hashCode);
+    FieldPropertyInfo fieldPropertyInfo = FIELD_CACHE.get(hashCode);
     if (fieldPropertyInfo == null) {
       fieldPropertyInfo = new FieldPropertyInfo(type, field, name);
-      cache.put(hashCode, fieldPropertyInfo);
+      FIELD_CACHE.put(hashCode, fieldPropertyInfo);
     }
 
     return fieldPropertyInfo;
@@ -89,10 +93,10 @@ class PropertyInfoRegistry {
   static synchronized Mutator mutatorFor(Class<?> type, Method method, Configuration configuration,
       String name) {
     Integer hashCode = hashCodeFor(type, method.getName(), configuration);
-    Mutator mutator = (Mutator) cache.get(hashCode);
+    Mutator mutator = MUTATOR_CACHE.get(hashCode);
     if (mutator == null) {
       mutator = new MethodMutator(type, method, name);
-      cache.put(hashCode, mutator);
+      MUTATOR_CACHE.put(hashCode, mutator);
     }
 
     return mutator;

--- a/core/src/test/java/org/modelmapper/bugs/GH204.java
+++ b/core/src/test/java/org/modelmapper/bugs/GH204.java
@@ -1,0 +1,116 @@
+package org.modelmapper.bugs;
+
+import static org.testng.Assert.assertEquals;
+
+import org.modelmapper.AbstractTest;
+import org.modelmapper.PropertyMap;
+import org.testng.annotations.Test;
+
+@Test
+public class GH204 extends AbstractTest {
+  static class SomeDto {
+    private String someValue;
+
+    public SomeDto() {
+    }
+
+    public SomeDto(String someValue) {
+      this.someValue = someValue;
+    }
+
+    public String someValue() {
+      return someValue;
+    }
+
+    public SomeDto someValue(String someValue) {
+      this.someValue = someValue;
+      return this;
+    }
+  }
+
+  static class SomeEntity {
+    private String otherValue;
+
+    public SomeEntity() {
+    }
+
+    public SomeEntity(String otherValue) {
+      this.otherValue = otherValue;
+    }
+
+    public String otherValue() {
+      return otherValue;
+    }
+
+    public SomeEntity otherValue(String otherValue) {
+      this.otherValue = otherValue;
+      return this;
+    }
+  }
+
+  public void shouldMappingMethod1() {
+    modelMapper.addMappings(new PropertyMap<SomeDto, SomeEntity>() {
+      @Override
+      protected void configure() {
+        map(source.someValue(), destination.otherValue(null));
+      }
+    });
+    modelMapper.addMappings(new PropertyMap<SomeEntity, SomeDto>() {
+      @Override
+      protected void configure() {
+        map(source.otherValue(), destination.someValue(null));
+      }
+    });
+
+    modelMapper.validate();
+
+    assertEquals("foo",
+        modelMapper.map(new SomeDto("foo"), SomeEntity.class).otherValue);
+    assertEquals("bar",
+        modelMapper.map(new SomeEntity("bar"), SomeDto.class).someValue);
+  }
+
+  public void shouldMappingMethod2() {
+    modelMapper.addMappings(new PropertyMap<SomeDto, SomeEntity>() {
+      @Override
+      protected void configure() {
+        map().otherValue(source.someValue());
+      }
+    });
+    modelMapper.addMappings(new PropertyMap<SomeEntity, SomeDto>() {
+      @Override
+      protected void configure() {
+        map().someValue(source.otherValue());
+      }
+    });
+
+    modelMapper.validate();
+
+    assertEquals("foo",
+        modelMapper.map(new SomeDto("foo"), SomeEntity.class).otherValue);
+    assertEquals("bar",
+        modelMapper.map(new SomeEntity("bar"), SomeDto.class).someValue);
+  }
+
+  public void shouldMappingMethod3() {
+    modelMapper.addMappings(new PropertyMap<SomeDto, SomeEntity>() {
+      @Override
+      protected void configure() {
+        map(source.someValue()).otherValue(null);
+      }
+    });
+    modelMapper.addMappings(new PropertyMap<SomeEntity, SomeDto>() {
+      @Override
+      protected void configure() {
+        map(source.otherValue()).someValue(null);
+      }
+    });
+
+    modelMapper.validate();
+
+    assertEquals("foo",
+        modelMapper.map(new SomeDto("foo"), SomeEntity.class).otherValue);
+    assertEquals("bar",
+        modelMapper.map(new SomeEntity("bar"), SomeDto.class).someValue);
+  }
+}

